# **Data Transformation**
# **Assignment**

**Q1. Handle Missing Values (Basic Cleaning)**

**Fill missing values using the following:**

**Age → Median**

**Salary → Mean**

**Hours_Worked_Per_Week → Median**

**Performance_Score → Mean**

**Display the dataset after handling missing values.**

In [5]:
import pandas as pd

# Load dataset
file = 'employee_productivity_dataset.csv'
df = pd.read_csv(file)

# Display first 5 rows
print(df.head())
print(df.info())


# Count missing values per column
print("count missing value",df.isnull().sum())

# Total missing values
print('Total missing:', df.isnull().sum().sum())

# Fill missing values
df['Age'].fillna(df['Age'].median())
df['Salary'].fillna(df['Salary'].mean())
df['Hours_Worked_Per_Week'].fillna(df['Hours_Worked_Per_Week'].median())
df['Performance_Score'].fillna(df['Performance_Score'].mean())

# Display cleaned dataset
print(df.head(12))


  Employee_ID         Name Department   Age  Gender Work_Mode   Join_Date  \
0        E001  Amit Sharma         IT  25.0    Male    Remote  15-01-2021   
1        E002   Riya Patel         HR  30.0  Female    Hybrid  15-02-2020   
2        E003  John Mathew    Finance   NaN    Male    Onsite  10-03-2022   
3        E004  Sneha Joshi         IT  28.0  Female    Remote  22-04-2021   
4        E005  Arjun Verma  Marketing  35.0    Male    Hybrid  18-05-2020   

   Hours_Worked_Per_Week  Projects_Completed  Performance_Score   Salary  \
0                   40.0                   5                4.5  50000.0   
1                   38.0                   4                4.0  45000.0   
2                   45.0                   6                NaN  60000.0   
3                    NaN                   5                4.2  52000.0   
4                   50.0                   7                4.8      NaN   

    Location  
0       Pune  
1     Mumbai  
2  Bangalore  
3      Delhi  
4    

**Q2. Label Encoding**

**Convert these columns into numeric values using Label Encoding:**

**Gender**

**Department**

**Show the updated columns**

In [6]:
from sklearn.preprocessing import LabelEncoder

# Initialize encoder
le_gender = LabelEncoder()
le_dept = LabelEncoder()

# Apply label encoding
df['Gender'] = le_gender.fit_transform(df['Gender'])
df['Department'] = le_dept.fit_transform(df['Department'])

# Show updated columns
print(df[['Gender', 'Department']].head(12))

# For reference, show mapping
print("Gender Mapping:", dict(zip(le_gender.classes_, le_gender.transform(le_gender.classes_))))
print("Department Mapping:", dict(zip(le_dept.classes_, le_dept.transform(le_dept.classes_))))


    Gender  Department
0        1           2
1        0           1
2        1           0
3        0           2
4        1           3
5        0           1
6        1           0
7        0           3
8        1           2
9        0           1
10       1           2
11       0           1
Gender Mapping: {'Female': np.int64(0), 'Male': np.int64(1)}
Department Mapping: {'Finance': np.int64(0), 'HR': np.int64(1), 'IT': np.int64(2), 'Marketing': np.int64(3)}


**Q3. One-Hot Encoding**

**Apply One-Hot Encoding on:**

**Work_Mode**

**Location**

**Display the dataset and check how many new columns are created.**

In [7]:
# One-Hot Encoding using pandas
df_encoded = pd.get_dummies(df, columns=['Work_Mode', 'Location'])

# Display first 12 rows
print(df_encoded.head(12))

# Check how many new columns were created
new_cols = [col for col in df_encoded.columns if 'Work_Mode_' in col or 'Location_' in col]
print("New Columns Created:", new_cols)
print("Total New Columns:", len(new_cols))


   Employee_ID         Name  Department   Age  Gender   Join_Date  \
0         E001  Amit Sharma           2  25.0       1  15-01-2021   
1         E002   Riya Patel           1  30.0       0  15-02-2020   
2         E003  John Mathew           0   NaN       1  10-03-2022   
3         E004  Sneha Joshi           2  28.0       0  22-04-2021   
4         E005  Arjun Verma           3  35.0       1  18-05-2020   
5         E006   Neha Singh           1  27.0       0  01-06-2023   
6         E007    Rahul Das           0  40.0       1  07-07-2019   
7         E008  Pooja Mehta           3  29.0       0  12-08-2022   
8         E009  Sameer Khan           2  31.0       1  30-09-2021   
9         E010   Kiran Nair           1  26.0       0  05-10-2023   
10        E001  Amit Sharma           2  25.0       1  15-01-2021   
11        E002   Riya Patel           1  30.0       0  15-02-2020   

    Hours_Worked_Per_Week  Projects_Completed  Performance_Score   Salary  \
0                    40.0

**Q4. Normalization (Min-Max Scaling)**

**Normalize the following columns:**

**Salary**

**Hours_Worked_Per_Week**

**Use MinMaxScaler and display results.**

In [8]:
from sklearn.preprocessing import MinMaxScaler

# Initialize scaler
scaler = MinMaxScaler()

# Apply Min-Max Scaling
df[['Salary', 'Hours_Worked_Per_Week']] = scaler.fit_transform(df[['Salary', 'Hours_Worked_Per_Week']])

# Display first 12 rows
print(df[['Salary', 'Hours_Worked_Per_Week']].head(12))


      Salary  Hours_Worked_Per_Week
0   0.347826               0.285714
1   0.130435               0.142857
2   0.782609               0.642857
3   0.434783                    NaN
4        NaN               1.000000
5   0.000000               0.000000
6   1.000000               0.857143
7   0.260870               0.428571
8   0.391304                    NaN
9   0.043478               0.071429
10  0.347826               0.285714
11  0.130435               0.142857


**Q5. Standardization (Scaling)**

**Apply StandardScaler on:**

**Age**
**Projects_Completed**

**Display the transformed values.**

In [9]:
from sklearn.preprocessing import StandardScaler

# Initialize scaler
scaler_std = StandardScaler()

# Apply Standardization
df[['Age', 'Projects_Completed']] = scaler_std.fit_transform(df[['Age', 'Projects_Completed']])

# Display first 12 rows
print(df[['Age', 'Projects_Completed']].head(12))


         Age  Projects_Completed
0  -1.140068           -0.211604
1  -0.024784           -0.916949
2        NaN            0.493742
3  -0.470898           -0.211604
4   1.090500            1.199088
5  -0.693954           -1.622295
6   2.205783            1.904433
7  -0.247841            0.493742
8   0.198273           -0.211604
9  -0.917011           -0.916949
10 -1.140068           -0.211604
11 -0.024784           -0.916949


**Q6. Compare Scaling Methods**

**Apply both:**

**MinMaxScaler**

**StandardScaler**

**On the Salary column.**

**Show both results side by side.**

In [10]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler
import pandas as pd

# Extract Salary column
salary = df[['Salary']].copy()

# Apply Min-Max Scaling
minmax_scaler = MinMaxScaler()
salary['Salary_MinMax'] = minmax_scaler.fit_transform(salary[['Salary']])

# Apply StandardScaler
std_scaler = StandardScaler()
salary['Salary_Standard'] = std_scaler.fit_transform(salary[['Salary']])

# Display first 12 rows
print(salary.head(12))


      Salary  Salary_MinMax  Salary_Standard
0   0.347826       0.347826        -0.092450
1   0.130435       0.130435        -0.785825
2   0.782609       0.782609         1.294300
3   0.434783       0.434783         0.184900
4        NaN            NaN              NaN
5   0.000000       0.000000        -1.201850
6   1.000000       1.000000         1.987676
7   0.260870       0.260870        -0.369800
8   0.391304       0.391304         0.046225
9   0.043478       0.043478        -1.063175
10  0.347826       0.347826        -0.092450
11  0.130435       0.130435        -0.785825


**Q7. Build Preprocessing Pipeline**

**Create a pipeline that:**

**Applies encoding to categorical columns**

**Applies scaling to numerical columns**

**Use:**

**ColumnTransformer**

**Pipeline**

In [13]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Define categorical and numerical columns
categorical_cols = ['Gender', 'Department', 'Work_Mode', 'Location']
numerical_cols = ['Age', 'Projects_Completed', 'Salary', 'Hours_Worked_Per_Week', 'Performance_Score']

# ColumnTransformer: OneHotEncoder for categorical, StandardScaler for numerical
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols),
        ('num', StandardScaler(), numerical_cols)
    ]
)

# Build pipeline
pipeline = Pipeline(steps=[('preprocessor', preprocessor)])

# Fit and transform dataset
df_processed = pipeline.fit_transform(df)

# Convert to DataFrame (no .toarray needed)
df_processed = pd.DataFrame(df_processed)

# Display first 12 rows
print(df_processed.head(12))
print("Transformed shape:", df_processed.shape)


     0    1    2    3    4    5    6    7    8    9    10   11   12        13  \
0   0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0 -1.140068   
1   1.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0 -0.024784   
2   0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0       NaN   
3   1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0 -0.470898   
4   0.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  1.090500   
5   1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0 -0.693954   
6   0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  2.205783   
7   1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0 -0.247841   
8   0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.198273   
9   1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0 -0.917011   
10  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0 -1.140068   
11  1.0  0.0  0.0  1.0  0.0 

**Q8. Apply Pipeline**

**Apply the pipeline on the dataset.**

**Display:**

**Transformed dataset**

**Shape of final dataset**

In [14]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Define categorical and numerical columns
categorical_cols = ['Gender', 'Department', 'Work_Mode', 'Location']
numerical_cols = ['Age', 'Projects_Completed', 'Salary', 'Hours_Worked_Per_Week', 'Performance_Score']

# ColumnTransformer: OneHotEncoder for categorical, StandardScaler for numerical
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols),
        ('num', StandardScaler(), numerical_cols)
    ]
)

# Build pipeline
pipeline = Pipeline(steps=[('preprocessor', preprocessor)])

# Fit and transform dataset
df_processed = pipeline.fit_transform(df)

# Convert to DataFrame
df_processed = pd.DataFrame(df_processed)

# Display transformed dataset (first 12 rows)
print(df_processed.head(12))

# Display shape of final dataset
print("Final dataset shape:", df_processed.shape)


     0    1    2    3    4    5    6    7    8    9    10   11   12        13  \
0   0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0 -1.140068   
1   1.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0 -0.024784   
2   0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0       NaN   
3   1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0 -0.470898   
4   0.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  1.090500   
5   1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0 -0.693954   
6   0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  2.205783   
7   1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0 -0.247841   
8   0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.198273   
9   1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0 -0.917011   
10  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0 -1.140068   
11  1.0  0.0  0.0  1.0  0.0 

**Q9. Conceptual Question**

**Why is scaling important in Python?**

**Ans:**  Scaling is crucial in Python because many machine learning algorithms are sensitive to the magnitude of features. Without scaling, models like SVMs, Logistic Regression, and KNN can perform poorly or converge slowly, while properly scaled data ensures fair feature contribution and faster, more accurate training.

Why Scaling Matters in Python
1. **Ensures Equal Feature Contribution**
Features like Salary (₹65,000) and Performance Score (4.5) exist on very different scales.

Without scaling, algorithms may give undue importance to large-scale features (like Salary) while ignoring smaller ones (like Performance Score).

Scaling ensures all features contribute equally.

2. **Improves Algorithm Performance**
Distance-based models (KNN, K-Means, PCA) rely on Euclidean distance. If one feature dominates due to scale, clustering or classification becomes biased.

Gradient-based models (Logistic Regression, Neural Networks) converge faster when features are standardized.

3. **Stabilizes Optimization**
Scaling reduces numerical instability during optimization.

For example, StandardScaler (Z-score normalization) makes features centered around mean = 0, std = 1, which helps gradient descent converge efficiently.

4. **Model-Specific Needs**
Tree-based models (Random Forest, XGBoost) are largely unaffected by scaling.

Linear models, SVMs, PCA, KNN require scaling for meaningful results.

Choosing the right scaler depends on the algorithm and dataset properties.

**Q10. Conceptual Question**

**Why do we convert categorical data into numerical form?**

**Ans:**  We convert categorical data into numerical form because most machine learning algorithms and statistical models work with numbers, not text. If we leave categories as strings (like Male/Female or HR/IT/Finance), the model cannot interpret them mathematically.

Reasons for Converting Categorical Data
1. **Model Compatibility**
Algorithms like Logistic Regression, SVM, KNN, Neural Networks require numerical inputs.

Text labels cannot be directly processed in mathematical equations.

2. **Mathematical Operations**
Models calculate distances, correlations, and weights.

For example, KNN uses Euclidean distance — it cannot compute distance between “HR” and “Finance” unless they are represented numerically.

3. **Avoiding Misinterpretation**
Assigning arbitrary numbers (HR = 1, IT = 2, Finance = 3) may imply an order that doesn’t exist.

Techniques like One-Hot Encoding or Label Encoding ensure categories are represented correctly without false hierarchy.

4. **Improves Accuracy**
Proper encoding allows the model to capture relationships between categorical features and target variables.

Example: Work_Mode (Remote/Hybrid/Onsite) encoded numerically helps predict productivity more effectively.

5. **Flexibility Across Algorithms**
Tree-based models (Random Forest, XGBoost) can handle categorical splits but still benefit from numerical encoding.

Linear models & distance-based models absolutely require numerical conversion.